File: make_hourly.r
# convert min/max weather format into hourly format
library(data.table)
library(lubridate)
source("util.r")

C_TEMP <- list(c_alpha = 0.2, c_beta = 2.0, c_gamma = -2.9)
C_RH <- list(c_alpha = 0.4, c_beta = 1.9, c_gamma = -2.9)
C_WIND <- list(c_alpha = 1.2, c_beta = 1.7, c_gamma = -1.5)

makePrediction <- function(fcsts, c_alpha, c_beta, c_gamma, v = "TEMP", change_at = "SUNSET", min_value = -Inf, max_value = Inf, intervals = 1, verbose = FALSE) {
  if (verbose) {
    print(paste0("Predicting ", v, " changing at ", change_at))
  }
  fcsts <- copy(fcsts)
  var_min <- paste0(v, "_MIN")
  var_max <- paste0(v, "_MAX")
  fcsts[, `:=`(
    TIME_MIN = SUNRISE + c_alpha,
    TIME_MAX = (SOLARNOON + c_beta),
    VAR_MIN_TOM = data.table::shift(get(var_min), -1),
    TIME_MIN_TOM = data.table::shift(SUNRISE, -1) + c_alpha
  )]
  # ~ fcsts[, `:=`(r2_time_min = data.table::shift(TIME_MIN, -1),
  # ~ r2_var_min = data.table::shift(get(var_min), -1))]
  stopifnot(1 == nrow(unique(fcsts$ID)))
  hours <- data.table(HR = 0:23)
  cross <- as.data.table(merge(as.data.frame(hours), as.data.frame(fcsts[, c("DATE")]), all = TRUE))
  minutes <- data.table(MINUTE = seq(0, 59, by = 60 / intervals))
  cross <- as.data.table(merge(as.data.frame(cross), as.data.frame(minutes), all = TRUE))
  cross[, ID := unique(fcsts$ID)]
  setorder(cross, cols = "DATE", "HR", "MINUTE")
  cross[, TIME := HR + MINUTE / 60.0]
  # cross <- merge(cross, fcsts[, -c('HR', 'MINUTE', 'TIME')], by=c('ID', 'DATE'))
  for (column in c("HR", "MINUTE", "TIME", "MON", "YR"))
  {
    if (column %in% colnames(fcsts)) {
      set(fcsts, , column, NULL)
    }
  }
  cross <- merge(cross, fcsts, by = c("ID", "DATE"))
  cross[, `:=`(
    START_DATE = ifelse(TIME <= TIME_MIN, as.character(as.Date(DATE) - 1), DATE),
    HOUR_CURVE = ifelse(TIME <= TIME_MIN, TIME + 24, TIME),
    IS_RISING = ifelse(TIME <= TIME_MIN, FALSE, TIME < get(change_at)),
    TIME_G_MIN = ifelse(TIME <= TIME_MIN, TIME_MIN, TIME_MIN_TOM),
    VAR_G_MIN = ifelse(TIME <= TIME_MIN, get(var_min), VAR_MIN_TOM)
  ), ]
  rising <- cross[IS_RISING == TRUE, ]
  falling <- cross[IS_RISING == FALSE, ]
  # figure out values before change_at
  rising[, F_OR_G := (HOUR_CURVE - TIME_MIN) / (TIME_MAX - TIME_MIN)]
  rising[, eval(v) := get(var_min) + (get(var_max) - get(var_min)) * sin((pi / 2) * F_OR_G)]
  rising[, VAR_CHANGE := get(var_min) + (get(var_max) - get(var_min)) * sin((pi / 2) * ((get(change_at) - TIME_MIN) / (TIME_MAX - TIME_MIN)))]
  # need to figure out var value at change_at
  tmp <- unique(rising[, c("START_DATE", "VAR_CHANGE")])
  # now merge change_at values into falling
  falling <- merge(falling, tmp, by = c("START_DATE"))
  falling[, F_OR_G := (HOUR_CURVE - get(change_at)) / (24 - get(change_at) + TIME_G_MIN)]
  falling[, eval(v) := VAR_G_MIN + (VAR_CHANGE - VAR_G_MIN) * exp(c_gamma * F_OR_G)]
  # combine and sort everything
  out <- rbind(rising, falling)
  setorder(out, cols = "DATE", "HR")
  out[, TIMESTAMP := as_datetime(paste0(as.character(DATE), " ", HR, ":", MINUTE, ":00"))]
  out[, eval(v) := pmin(pmax(get(v), min_value), max_value)]
  # HACK: somehow this isn't returning unless we do something to it first
  out <- out[!is.na(get(v))]
  return(copy(out))
}


doPrediction <- function(fcsts, row_temp, row_wind, row_RH, intervals = 1, verbose = FALSE) {
  if (verbose) {
    print("Doing prediction")
  }
  v_temp <- makePrediction(fcsts, row_temp$c_alpha, row_temp$c_beta, row_temp$c_gamma, "TEMP", "SUNSET", intervals = intervals, verbose = verbose)
  v_wind <- makePrediction(fcsts, row_wind$c_alpha, row_wind$c_beta, row_wind$c_gamma, "WS", "SUNSET", min_value = 0, intervals = intervals, verbose = verbose)
  t <- v_temp[, c("ID", "TIMESTAMP", "DATE", "HR", "LAT", "LONG", "TEMP")]
  w <- v_wind[, c("ID", "TIMESTAMP", "DATE", "HR", "WS")]
  out <- merge(t, w)
  RH <- makePrediction(fcsts, row_RH$c_alpha, row_RH$c_beta, row_RH$c_gamma, "RH_OPP", intervals = intervals, min_value = 0, max_value = 1, verbose = verbose)
  RH[, `:=`(RH = 100 * (1 - RH_OPP))]
  RH <- RH[, c("ID", "TIMESTAMP", "RH")]
  out <- merge(RH, out, by = c("ID", "TIMESTAMP"))
  output <- out[, c("ID", "LAT", "LONG", "TIMESTAMP", "TEMP", "WS", "RH")]
  # ~ output <- fwi(output)
  if (verbose) {
    print("Assigning times")
  }
  output[, HR := hour(TIMESTAMP)]
  output[, MINUTE := minute(TIMESTAMP)]
  if (verbose) {
    print("Converting date")
  }
  output[, DATE := as.character(as_date(TIMESTAMP))]
  if (verbose) {
    print("Allocating rain")
  }
  prec <- fcsts[, c("DATE", "PREC")]
  prec[, HR := 7]
  prec[, MINUTE := 0]
  if (verbose) {
    print("Merging")
  }
  cmp <- merge(output, prec, by = c("DATE", "HR", "MINUTE"), all = TRUE)[!is.na(TIMESTAMP)]
  cmp$PREC <- nafill(cmp$PREC, fill = 0)
  if (verbose) {
    print("Calculating times")
  }
  cmp[, YR := year(TIMESTAMP)]
  cmp[, MON := month(TIMESTAMP)]
  cmp[, DAY := day(TIMESTAMP)]
  cmp[, TIME := HR + MINUTE / 60.0]
  if (verbose) {
    print("Done prediction")
  }
  return(cmp)
}


#' Convert daily min/max values stream to hourly values stream.
#' Uses Beck & Trevitt method with default A/B/G values.
#'
#' @param     w         daily min/max values weather stream [lat, long, yr, mon, day, hr, temp_min, temp_max, rh_min, rh_max, ws_min, ws_max, prec]
#' @param     timezone  integer offset from GMT to use for sun calculations
#' @return              hourly values weather stream [lat, long, yr, mon, day, hr, temp, rh, ws, prec]
minmax_to_hourly_single <- function(w, timezone, skipInvalid = FALSE, verbose = FALSE) {
  r <- copy(w)
  colnames(r) <- toupper(colnames(r))
  if (length(na.omit(unique(r$LAT))) != 1) {
    stop("Expected a single LAT value for input weather")
  }
  if (length(na.omit(unique(r$LONG))) != 1) {
    stop("Expected a single LONG value for input weather")
  }
  hadId <- TRUE
  if (!("ID" %in% colnames(r))) {
    r$ID <- 1
    hadId <- FALSE
  } else if (length(na.omit(unique(r$ID))) != 1) {
    stop("Expected a single ID value for input weather")
  }
  if (length(na.omit(unique(r$YR))) != 1) {
    stop("Expected a single YR value for input weather")
  }
  r$HR <- 12
  r[, TIMESTAMP := as_datetime(sprintf("%04d-%02d-%02d %02d:%02d:00", YR, MON, DAY, HR, 0),
    tz = paste0("Etc/GMT", ifelse(timezone > 0, "-", "+"), abs(timezone))
  )]
  if (!(nrow(r) > 1 && isSequential(as.Date(r$TIMESTAMP)))) {
    if (skipInvalid) {
      warning(paste0(r$ID[[1]], " for ", r$YR[[1]], " - Expected input to be sequential daily weather"))
      return(NULL)
    }
    stop("Expected input to be sequential daily weather")
  }
  orig_dates <- data.table(date = as.character(unique(as_date(r$TIMESTAMP))))
  # duplicate start and end dates so we can use their values for yesterday and tomorrow in predictions
  yest <- r[1, ]
  tom <- r[nrow(r), ]
  yest[, TIMESTAMP := TIMESTAMP - days(1)]
  tom[, TIMESTAMP := TIMESTAMP + days(1)]
  r <- rbind(yest, r, tom)
  r[, DATE := as_date(TIMESTAMP)]
  r[, YR := year(TIMESTAMP)]
  r[, MON := month(TIMESTAMP)]
  r[, DAY := day(TIMESTAMP)]
  r[, HR := hour(TIMESTAMP)]
  # FIX: convert this to not need to do individual stations
  r[, TIMEZONE := timezone]
  r <- getSunlight(r, with_solrad = FALSE)
  # FIX: is solar noon just midpoint between sunrise and sunset?
  r[, SOLARNOON := (SUNSET - SUNRISE) / 2 + SUNRISE]
  r[, RH_OPP_MIN := 1 - RH_MAX / 100]
  r[, RH_OPP_MAX := 1 - RH_MIN / 100]
  r[, DATE := as.character(DATE)]
  df <- doPrediction(r, row_temp = C_TEMP, row_wind = C_WIND, row_RH = C_RH, verbose = verbose)
  colnames(df) <- tolower(colnames(df))
  df <- merge(orig_dates, df, by = "date")
  if (hadId) {
    df <- df[, c("id", "lat", "long", "yr", "mon", "day", "hr", "temp", "rh", "ws", "prec")]
  } else {
    df <- df[, c("lat", "long", "yr", "mon", "day", "hr", "temp", "rh", "ws", "prec")]
  }
  return(df)
}

#' Convert daily min/max values stream to hourly values stream.
#' Uses Beck & Trevitt method with default A/B/G values.
#'
#' @param     w         daily min/max values weather stream [id, lat, long, yr, mon, day, hr, temp_min, temp_max, rh_min, rh_max, ws_min, ws_max, prec]
#' @param     timezone  integer offset from GMT to use for sun calculations
#' @return              hourly values weather stream [id, lat, long, yr, mon, day, hr, temp, rh, ws, prec]
#' @export minmax_to_hourly
minmax_to_hourly <- function(w, timezone, skipInvalid = FALSE, verbose = FALSE) {
  r <- copy(w)
  colnames(r) <- toupper(colnames(r))
  hadId <- TRUE
  if (!("ID" %in% colnames(r))) {
    r$ID <- 1
    hadId <- FALSE
  }
  result <- NULL
  for (stn in unique(r$ID))
  {
    by_stn <- r[ID == stn, ]
    for (yr in unique(by_stn$YR))
    {
      by_year <- by_stn[YR == yr, ]
      print(paste0("Running ", stn, " for ", yr))
      df <- minmax_to_hourly_single(by_year, timezone, skipInvalid, verbose)
      result <- rbind(result, df)
    }
  }
  if (!hadId) {
    result <- result[, c("lat", "long", "yr", "mon", "day", "hr", "temp", "rh", "ws", "prec")]
  }
  return(result)
}

# so this can be run via Rscript
if ("--args" %in% commandArgs()) {
  args <- commandArgs(trailingOnly = TRUE)
  if (3 == length(args)) {
    timezone <- as.double(args[1])
    inp <- args[2]
    out <- args[3]
    df <- as.data.table(read.csv(inp))
    df_hourly <- minmax_to_hourly(df, timezone)
    save_csv(df_hourly, out)
  } else {
    message("Wrong number of arguments")
  }
}

File: make_hourly.c
/*
inputs min/max weather stream
outputs hourly weather stream
*/

#include <stdio.h>
#include <math.h>
#include <float.h>
#include <stdlib.h>
#include <string.h>
#include "util.h"

/* Alpha/Beta/Gamma coefficients for Temperature, RH, and Wind Speed */
const float C_TEMP[3] = {0.3, 2.0, -3.3};
const float C_RH[3] = {0.4, 2.0, -3.4};
const float C_WIND[3] = {0.3, 3.5, -3.3};

/* A row from the input file */
struct row {
  float lat, lon;
  int year, mon, day, hour;
  float temp_min, temp_max, rh_min, rh_max, wind_min, wind_max, rain;
};

float diurnal(float v_min, float v_max, float tv_min, float yv_sunset, float sunrise, float sunset, float solarnoon_yest, float sunset_yest, float sunrise_tom, const float* abg, float* hourly);

int read_row(FILE *inp, struct row* r)
{
  char a[1]; /* this is declared as an array just to make it a pointer ...for reading commas easily*/
  return fscanf(inp,"%f%c%f%c%d%c%d%c%d%c%d%c%f%c%f%c%f%c%f%c%f%c%f%c%f",
  &r->lat,a,&r->lon,a,&r->year,a,&r->mon,a,&r->day,a,&r->hour,a,&r->temp_min,a,&r->temp_max,a,&r->rh_min,a,&r->rh_max,a,&r->wind_min,a,&r->wind_max,a,&r->rain);
}

void main(int argc, char *argv[]){
  const char* header = "lat,long,year,mon,day,hour,temp_min,temp_max,rh_min,rh_max,wind_min,wind_max,rain";
  FILE *inp, *out;
  int err;

  float TZadjust;
  char a[1]; /* this is declared as an array just to make it a pointer ...for reading commas easily*/

   if(argc != 4){
       printf("Command line:   %s <local GMToffset> <input file> <output file>\n\n", argv[0]);
       printf("<local GMToffset> is the off of Greenich mean time (for Eastern = -5  Central=-6   MT=-7  PT=-8 )  \n");
       printf("INPUT FILE format must be DAILY min/max weather data, comma seperated and take the form\n");
       printf("%s\n\n", header);
       printf("All times should be local standard time\n");
       exit(1);
   }

   inp=fopen(argv[2],"r");
   printf("Openning input file >>> %s   \n",argv[2]);
   if(inp==NULL){printf("/n/n ***** FILE  %s  does not exist\n",argv[2]);exit(1);}
   out=fopen(argv[3],"w");

   TZadjust=atoi(argv[1]);
   if(TZadjust <-9 || TZadjust> -2 ){ printf("/n *****   LOCal time zone adjustment must be vaguely in CAnada so between -9 and -2 \n"); exit(1);}

  check_header(inp, header);
  /*  CSV headers */
   fprintf(out, "%s\n", "lat,long,year,mon,day,hour,temp,rh,wind,rain");

   struct row tom;
   struct row cur;
   struct row yday;
   
   err=read_row(inp, &cur);
   if (!(err > 0))
   {
        printf("Unable to read first row of indices\n");
        exit(1);
   }
   /* use today's values for yesterday because we need to do something */
   yday = cur;
   float solarnoon_yest = 0.0;
   float sunset_yest = 0.0;
   float ytemp_sunset = 0.0;
   float yrh_sunset = 0.0;
   float ywind_sunset = 0.0;
   int done_first = 0;
   while(err>0){
     if (12 != cur.hour)
     {
       printf("Expected daily weather (hour value should be 12 but got %d)\n", cur.hour);
       exit(1);
     }
     if (done_first)
     {
       err = read_row(inp, &tom);
     } else {
       yday = cur;
       tom = cur;
     }
     if (!(err > 0))
     {
       /* error but we still want to figure out today's values */
       tom = cur;
     }
     printf("%d %d %d %d  %5.1f  %5.1f %5.1f %5.1f %5.1f %5.1f %5.1f\n",
            cur.year, cur.mon, cur.day, cur.hour,
            cur.temp_min, cur.temp_max,
            cur.rh_min, cur.rh_max,
            cur.wind_min, cur.wind_max,
            cur.rain);
     int jd = julian(cur.mon, cur.day);
     float sunrise, sunset;
     float solar=sun_julian(cur.lat,cur.lon,jd,cur.hour,TZadjust,&sunrise,&sunset );
     float solarnoon = (sunset - sunrise) / 2.0 + sunrise;
     float sunrise_tom, sunset_tom;
     float solar_tom=sun_julian(cur.lat,cur.lon,jd+1,cur.hour,TZadjust,&sunrise_tom,&sunset_tom );
     float atemp[24], arh[24], awind[24], arain[24];
     int h;
     for (h = 0; h < 24; ++h)
     {
       atemp[h] = 0.0;
       arh[h] = 0.0;
       awind[h] = 0.0;
       arain[h] = 0.0;
     }
     ytemp_sunset = diurnal(cur.temp_min, cur.temp_max, tom.temp_min, ytemp_sunset, sunrise, sunset, solarnoon_yest, sunset_yest, sunrise_tom, C_TEMP, atemp);
     yrh_sunset = diurnal(1.0 - cur.rh_max/100.0, 1.0-cur.rh_min/100.0, 1.0-tom.rh_max/100.0, yrh_sunset, sunrise, sunset, solarnoon_yest, sunset_yest, sunrise_tom, C_RH, arh);
     ywind_sunset = diurnal(cur.wind_min, cur.wind_max, tom.wind_min, ywind_sunset, sunrise, sunset, solarnoon_yest, sunset_yest, sunrise_tom, C_WIND, awind);
     arain[7] = cur.rain;
     if (done_first)
     {
       for (h = 0; h < 24; ++h)
       {
          arh[h] = 100 * (1.0 - arh[h]);
          fprintf(out,"%.4f,%.4f,%4d,%02d,%02d,%02d,%.1f,%.0f,%.1f,%.1f\n",
                  cur.lat,cur.lon,cur.year,cur.mon,cur.day,h,atemp[h],arh[h],awind[h],arain[h]);
       }
     } else {
       done_first = 1;
     }
     /* move today to yesterday's variables */
     yday = cur;
     solarnoon_yest = solarnoon;
     sunset_yest = sunset;
     /* advance to the next day */
     cur = tom;
   }  /* end the main while(err>0)  */

  /* printf("output has been written to>>> %s\n",argv[6]); */
  fclose(inp);
  fclose(out);
}

/**
 * Determines values based on a diurnal curve
 *
 * @param v_min               minimum value today
 * @param v_max               maximum value today
 * @param tv_min              tomorrow's minimum value
 * @param yv_sunset           yesterday's value at sunset
 * @param sunrise             time of sunrise today
 * @param sunset              time of sunset today
 * @param solarnoon_yest      time of solar noon yesterday
 * @param sunset_yest         time of sunset yesterday
 * @param sunrise_tom         time of sunrise tomorrow
 * @param abg[3]              alpha, beta, gamma values to use
 * @param hourly[24]          array of calculated output values by hour for today
 * @return                    value at sunset today
 */
float diurnal(float v_min, float v_max, float tv_min, float yv_sunset, float sunrise, float sunset,
              float solarnoon_yest, float sunset_yest, float sunrise_tom, const float* abg, float* hourly)
{
  const float c_alpha = abg[0];
  const float c_beta = abg[1];
  const float c_gamma = abg[2];
  const float solarnoon = (sunset - sunrise) / 2.0 + sunrise;
  const float t_min = sunrise + c_alpha;
  const float t_max = solarnoon + c_beta;
  const float v_sunset = v_min + (v_max - v_min) * sin((pi / 2.0)*((sunset - t_min)/(t_max - t_min)));
  const float yt_min = solarnoon_yest + c_beta;
  const float time_min_tom = sunrise_tom + c_alpha;
  int h;
  for (h = 0; h < 24; ++h)
  {
    if (h <= t_min)
    {
      hourly[h] = v_min + (yv_sunset - v_min) * exp(c_gamma * ((h + 24 - sunset_yest)/(24 - sunset_yest + t_min)));
    } else if (h > t_min && h <= sunset) {
      hourly[h] = v_min + (v_max - v_min) * sin((pi / 2.0)*((h - t_min)/(t_max - t_min)));
    } else {
      /* h > sunset && h < 24 + t_min tomorrow */
      hourly[h] = tv_min + (v_sunset - tv_min) * exp(c_gamma * ((h - sunset)/(24 - sunset + time_min_tom)));
    }
  }
  return v_sunset;
}
